src/range_proof/mod.rs

```diff
@@ -137,17 +137,118 @@ impl RangeProof {
         transcript: &mut ProofTranscript,
         rng: &mut R,
         n: usize,
-    ) -> Result<(), ()> {
+    ) -> Result<(), &'static str> {
+        RangeProof::verify_batch(
+            iter::once((self, value_commitments, n)),
+            gens,
+            transcript,
+            rng,
+        )
+    }
+
+    /// Verifies multiple range proofs at once.
+    /// If any range proof is invalid, the whole batch is invalid.
+    /// Proofs may use different ranges (`n`) or different number of aggregated commitments (`m`).
+    /// You must provide big enough view into generators (`gens`) that covers
+    /// the biggest proof
+    pub fn verify_batch<'a,'b,I,R,P,V>(
+        proofs: I,
+        gens: GeneratorsView, // must have enough points to cover max(m*n)
+        transcript: &mut ProofTranscript,
+        rng: &mut R
+    ) -> Result<(), &'static str> 
+    where
+    R: Rng,
+    I: IntoIterator<Item = (P, V, usize)>,
+    P: Borrow<RangeProof>,
+    V: AsRef<[RistrettoPoint]>
+    {
+        let mut nm: usize = 0;
+        let batch = proofs.into_iter().map(|(p, vcs, n)| {
+            let m = vcs.as_ref().len();
+            let v = p.borrow().prepare_verification(n, vcs, &mut transcript.clone(), rng);
+            nm = nm.max(n*m);
+            v
+        }).collect::<Vec<_>>();
+
+        if gens.G.len() < nm {
+            return Err(
+                "The generators view does not have enough generators for the largest proof",
+            );
+        }
+
+        // First statement is used without a random factor
```

Comment:
IMO it is cleaner not to unroll the first loop iteration, and just multiply all statements by a random factor. The code is simpler, the additional cost is minimal (and shrinks as the batch size grows).